### ISSUE TITLE
Deleting outDir breaks Linaria for good

### ISSUE THREAD

**Issue by steadicat** [callstack/linaria#169]
**Do you want to request a *feature* or report a *bug*?**

Bug.

**What is the current behavior?**

I sometimes see stale generated stylesheets (more on that perhaps in future issues), so I find myself deleting `./src/.linaria-cache` (my `outDir`) to try to start fresh. After I do this, Linaria fails to build, spewing a bunch of errors like:

```
`Error in ./src/ui/components/chrome/SiteServicesMenu.js
Module not found: [CaseSensitivePathsPlugin] `src/.linaria-cache/src/ui/components/chrome/SiteServicesMenu.css` does not match the corresponding path on disk - File does not exist.
**If the current behavior is a bug, please provide the steps to reproduce and a minimal repository on GitHub that we can `yarn install` and `yarn test`.**
```

Restarting Webpack or clearing the Babel cache does not help. I'm stuck.

**What is the expected behavior?**

Missing output files get rebuilt instead of throwing errors.

**Please provide your exact Babel configuration and mention your Linaria, Node, Yarn/npm version and operating system.**

This is with `create-react-app`, Node 8.9.0, Webpack 1.14.0, Babel 6.2.7. This is how I configured Babel:

```
{
    "presets": [
      "react-app",
      [
        "linaria/babel",
        {
          "outDir": "./src/.linaria-cache"
        }
      ]
    ]
  }
```

**Comment by satya164** [comment]
I can successfully reproduce this. This happens due to `babel-loader`'s cache and gets fixed if you delete `node_modules/.cache/babel-loader`.

**Comment by steadicat** [comment]
I actually tried deleting the Babel Loader cache and it didn't seem to work. Deleting the whole `node_modules` directory did work for me though.

**Comment by satya164** [comment]
Hmm that's weird. Maybe you have some other cache too? I tried with ejecting a fresh CRA project and deleting that folder did the trick for me.

**Comment by zamotany** [comment]
Which linaria version are you using?

**Comment by steadicat** [comment]
I can repro with both `0.2.0` and `0.3.1`.

**Comment by steadicat** [comment]
This happens also when you start fresh with a clean repo, and no `outDir` folder. Deleting `node_modules` doesn't seem to help either. This is bad. Is there a workaround? Do we need to touch all the source files or something?

**Comment by thymikee** [comment]
Does building whole project helps? This would restore whole `outDir`.

**Comment by steadicat** [comment]
I'll try that, thanks. Building the project takes several minutes though (tons of JS minification and image optimization). Is there a better way? How does this normally work on other projects?

**Comment by thymikee** [comment]
We just didn't encounter this in our testing, but that's definitely something buggy which we need to fix. 
Other than that, building the project doesn't mean you need to run in production mode, it's all good with dev (which should be optimised for speed, so no minification et. al.). What's important here is to get all files processed by Babel (which runs Linaria). You can also try `cache: false` in the preset options: https://github.com/callstack/linaria/blob/master/docs/BABEL_PRESET.md#options because it might be related

**Comment by zamotany** [comment]
`cache: false` would only make difference between rebuilds in dev mode (same JS context), if you build your project, spawning a new command every time, linaria cache will always be empty and files will be created.

**Comment by zamotany** [comment]
I've tried to reproduce this issue within `website` project with no luck. It would greatly help you you could create a repo when we can reproduce this bug.

**Comment by satya164** [comment]
@zamotany I tried with a fresh project and that doesn't happen even if I restart webpack. The babel plugin never gets called when `cacheDirectory` is specified for `babel-loader`. The only way to fix this is to delete `node_modules/.cache/babel-loader`

**Comment by steadicat** [comment]
Here's a [repo where you can reproduce it](/steadicat/linaria-test). I simply created an app with `create-react-app` and ejected to add Linaria. Run it, delete `src/.linaria-cache`, and you'll see this error:
<img width="821" alt="screen shot 2017-11-13 at 7 49 14 pm" src="https://user-images.githubusercontent.com/6031/32762052-e6959c0c-c8ab-11e7-8c5b-2447a7a0740c.png">

You’ll also run into this bug if you pull someone else’s changes. You’ll get stale CSS, or a completely broken build, until you run `yarn build` or delete the Babel cache. This makes it really inconvenient to collaborate on a Linaria project.

**Comment by steadicat** [comment]
I thought about this more and I realized that I might be conflating two separate issues:

1. If you delete `outDir` and not `node_modules/.cache/babel-loader`, generated CSS is missing. This is easily explained by Babel not running on unchanged and previously cached files. Deleting `node_modules/.cache/babel-loader` fixes it, as does disabling the Babel loader cache. Or simply avoid deleting `outDir`: this bug should never repro during normal usage.
2. In our repo, generated CSS is often missing or stale. Typically this happens when you pull someone else’s changes. Deleting `node_modules/.cache/babel-loader` **does not fix this**.  Disabling the Babel Loader cache **does not fix this**. I can repro this consistently in our repo by running `rm src/.linaria-cache && rm node_modules/.cache/babel-loader && yarn start`. Unfortunately I cannot repro in the simpler [test repo](/steadicat/linaria-test). This is *not* explained by Babel not running on unchanged files, because files do change when you pull. Issue 1 should also never result in stale CSS, which we are seeing consistently.

2 is the more pressing issue for us, and the one that really has me stumped. I have no idea where state is being stored between builds. Let me know if I should open a separate issue, or if there's anything I can do to help track this down.

**Comment by zamotany** [comment]
It theory or at least when we were developing website we have not encountered stale/missing CSS files. Files stored in `.linaria-cache` or in `outDir` are only there, so that the bundler can include then the bundle and they are regenerated every time (as long as some caching mechanism is not interfering). So rebuilding the project should emit a fresh and up-to-date CSS files. With all that said, without more context on how you build your project, do you rebuild when pulling changes, which files are modified and where they are used, I don't really know how to help you.

Even after fixing the first problem, it won't solve stale/missing CSS.

It is possible that what you described as stale/missing CSS is situation when you change `moduleA.js` used in `moduleB.js` which emits a CSS file? Sth like:
```js
// moduleA.js
ezport default { /* ... */ }; // later this is changed
```

```js
// moduleB.js
import moduleA from './moduleA';
const myVar = css`
  some-property: ${moduleA.someProperty}
`;
```
??

**Comment by steadicat** [comment]
Upgrading our dependencies seems to fix the worst of this. The remaining issue about deleting `outDir` without clearing the Babel cache is still there, but nowhere near as pressing.

**Comment by zamotany** [comment]
We released Webpack loader in `0.5.0`. It should solve all of your problems, so please git a try. I'm going to close this issue. Should you have any problem, feel free to reopen it or create a new one.
